From 991af78844630a3fd5721251282073a02c7eca8c Mon Sep 17 00:00:00 2001
From: Linjie Fu <linjie.fu@intel.com>
Date: Thu, 2 Jan 2020 11:12:15 +0800
Subject: [PATCH 15/94] lavc/decode: Add internal surface re-allocate method
 for hwaccel

Add HWACCEL_CAP_INTERNAL_ALLOC flag to indicate hwaccels are able to
re-allocate surface internally through ff_decode_get_hw_frames_ctx.

Signed-off-by: Linjie Fu <linjie.fu@intel.com>
---
 libavcodec/decode.c   | 36 ++++++++++++++++++++++++++++++++++++
 libavcodec/hwconfig.h |  1 +
 2 files changed, 37 insertions(+)

diff --git a/libavcodec/decode.c b/libavcodec/decode.c
index 5c8442a0ec..be1f51466d 100644
--- a/libavcodec/decode.c
+++ b/libavcodec/decode.c
@@ -1103,6 +1103,33 @@ static const AVCodecHWConfigInternal *get_hw_config(AVCodecContext *avctx, enum
     return hw_config;
 }
 
+static int hwaccel_realloc_surface(AVCodecContext *avctx)
+{
+    const AVCodecHWConfigInternal *hw_config;
+    int ret;
+
+    if (avctx->hw_frames_ctx)
+        av_buffer_unref(&avctx->hw_frames_ctx);
+
+    hw_config = get_hw_config(avctx, avctx->pix_fmt);
+    if (!hw_config)
+        return AV_PIX_FMT_NONE;
+
+    if (avctx->hw_device_ctx &&
+        hw_config->public.methods & AV_CODEC_HW_CONFIG_METHOD_HW_DEVICE_CTX) {
+        const AVHWDeviceContext *device_ctx =
+                (AVHWDeviceContext*)avctx->hw_device_ctx->data;
+        ret = ff_decode_get_hw_frames_ctx(avctx, device_ctx->type);
+        if (ret < 0) {
+            av_log(avctx, AV_LOG_WARNING, "Failed to re-allocate hwaccel surface internally.\n");
+            return AV_PIX_FMT_NONE;
+        }
+    } else
+        return AV_PIX_FMT_NONE;
+
+    return hw_config->public.pix_fmt;
+}
+
 int ff_get_format(AVCodecContext *avctx, const enum AVPixelFormat *fmt)
 {
     const AVPixFmtDescriptor *desc;
@@ -1129,6 +1156,15 @@ int ff_get_format(AVCodecContext *avctx, const enum AVPixelFormat *fmt)
         return AV_PIX_FMT_NONE;
 
     for (;;) {
+        if (avctx->internal->hwaccel_priv_data &&
+            avctx->hwaccel->caps_internal & HWACCEL_CAP_INTERNAL_ALLOC) {
+            err = hwaccel_realloc_surface(avctx);
+            if (err < 0)
+                av_log(avctx, AV_LOG_WARNING, "Try to re-initialize all.\n");
+            else
+                return err;
+        }
+
         // Remove the previous hwaccel, if there was one.
         hwaccel_uninit(avctx);
 
diff --git a/libavcodec/hwconfig.h b/libavcodec/hwconfig.h
index f421dc909f..6b2d0dde7a 100644
--- a/libavcodec/hwconfig.h
+++ b/libavcodec/hwconfig.h
@@ -24,6 +24,7 @@
 
 
 #define HWACCEL_CAP_ASYNC_SAFE      (1 << 0)
+#define HWACCEL_CAP_INTERNAL_ALLOC  (1 << 1)
 
 
 typedef struct AVCodecHWConfigInternal {
-- 
2.25.1

